packages/saber-plugin-amp/lib/index.js

```diff
@@ -0,0 +1,38 @@
+const path = require('path')
+
+const ID = 'amp'
+
+exports.name = ID
+
+exports.apply = api => {
+  api.hooks.afterExportPage.tapPromise(ID, async (context, exportedPage) => {
+    if (context.amp) {
+      const { log } = api
+      const { fs } = api.utils
+      const outputDir = api.resolveCache('public')
+
+      // amp-script is experimental for now
+      // this should change when it becomes available for production
+      context.renderScripts = () => ''
+
+      api.hooks.getDocumentData.tap(ID, data => {
+        data.htmlAttrs += ' amp'
+        data.link += `<link rel="canonical" href="${context.url}">`
+        data.style =
+          data.style.replace('<style', '<style amp-custom') +
+          `<style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style>`
+        data.noscript += `<noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>`
+      })
```

Comment:
@egoist is there any way to stop `api.hooks.getDocumentData.tap` here because it will get re-executed after each `afterExportPage` event (if there is more than one AMP page) and it's necessary to have it re-executed because of `context.url`

Reply:
No problem found a work around